packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts

```diff
@@ -0,0 +1,199 @@
+        resolve();
+      }
+    });
+  }
+}
+
+type DelegatedSignInIdentityOptions = {
+  provider: string;
+  discoveryApi: typeof discoveryApiRef.T;
+  fetchApi: typeof fetchApiRef.T;
+  errorApi: typeof errorApiRef.T;
+};
+
+/**
+ * An identity API that keeps itself up to date solely based on getting session
+ * information from a `/refresh` endpoint.
+ */
+export class DelegatedSignInIdentity implements IdentityApi {
+  private readonly options: DelegatedSignInIdentityOptions;
+  private readonly abortController: AbortController;
+  private session: DelegatedSession | undefined;
+
+  constructor(options: DelegatedSignInIdentityOptions) {
+    this.options = options;
+    this.abortController = new AbortController();
+    this.session = undefined;
+  }
+
+  async start() {
+    let signalErrors = true;
+    let pause = 0;
+
+    const refreshOnce = async () => {
+      if (!this.abortController.signal.aborted) {
+        try {
+          await this.refresh(signalErrors);
+          signalErrors = true;
+          pause = tokenToExpiryMillis(this.session?.backstageIdentity.token);
+          if (pause <= 0) {
+            this.session = undefined;
+          }
+        } catch {
+          signalErrors = false; // Only signal first failure in a row
+          pause = DEFAULTS.retryRefreshFrequencyMillis;
+        }
+      }
+    };
+
+    const refreshLoop = async () => {
+      while (!this.abortController.signal.aborted) {
+        await sleep(pause, this.abortController.signal);
+        await refreshOnce();
+      }
+    };
+
+    // Try first refresh and bubble up any errors to the caller, then get the
+    // loop going behind the scenes and return
+    await refreshOnce();
+    refreshLoop();
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getUserId} */
+  getUserId(): string {
+    return this.getSession().backstageIdentity.id;
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getIdToken} */
+  async getIdToken(): Promise<string | undefined> {
+    return this.getSession().backstageIdentity.token;
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getProfile} */
+  getProfile(): ProfileInfo {
+    return this.getSession().profile;
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getProfileInfo} */
+  async getProfileInfo(): Promise<ProfileInfo> {
+    return this.getSession().profile;
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getBackstageIdentity} */
+  async getBackstageIdentity(): Promise<BackstageUserIdentity> {
+    return this.getSession().backstageIdentity.identity;
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.getCredentials} */
+  async getCredentials(): Promise<{ token?: string | undefined }> {
+    return { token: this.getSession().backstageIdentity.token };
+  }
+
+  /** {@inheritdoc @backstage/core-plugin-api#IdentityApi.signOut} */
+  async signOut(): Promise<void> {
+    this.abortController.abort();
```

Comment:
I wasn't sure what to actually do here. Should we just throw an exception? Should we refresh the page? Should we for real try to sign out of the session somehow?